File: leetcode/2006-CountNumberOfPairsWithAbsoluteDifferenceK/solution.go
// Compute Complexity O(N)
// Space Complexity O(N)

func CalculateCount(hist map[int]int, target int) int {
	if targetCnt, ok := hist[target]; ok {
		return targetCnt
	}
	return 0
}

func countKDifference(nums []int, k int) int {
hist := make(map[int]int, 0)
counts := 0
for _, val := range nums {
	// find numbers that x - val = target => x = target + val
	counts += CalculateCount(hist, k + val)
	// find numbers that -x + val = target => x = val - target
	counts += CalculateCount(hist, val - k)              
	hist[val]++
}

return counts;
}

File: leetcode/2006-CountNumberOfPairsWithAbsoluteDifferenceK/solution.cpp
// Compute Complexity = O(N)
// Space Complexity = O(N)

class Solution {
private:
    int CalculateCount(const unordered_map<int, int>& hist, int target) {
        const auto it = hist.find(target);
        if (hist.end() != it) {
            return it->second;
        }
        return 0;
    }

public:
    int countKDifference(vector<int>& nums, int k) {
        auto count = 0;
        unordered_map<int, int> hist;

        for (const auto& val : nums) {
            // x - val = target => x = target + val
            count += CalculateCount(hist, k + val);

            // find total number of numbers val such that
            // -(x - val) = target => x = val - target
            count += CalculateCount(hist, val - k);

            hist[val]++;
        }
        return count;
    }
};